src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,185 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> previous;
+
+        Node(T value) {
+            this.value = value;
+        }
+
+        Node(Node<T> previous, Node<T> next, T value) {
+            this.previous = previous;
+            this.next = next;
+            this.value = value;
+        }
+
+    }
+
+    public MyLinkedList() {
+
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of range");
+        }
+        if (index == size) {
+            addTail(value);
+        } else {
+            addByIndex(value, index);
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list == null || list.isEmpty()) {
+            return false;
+        }
+        for (T t : list) {
+            add(t);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getCurrentNode(index);
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue;
+        Node<T> current = getCurrentNode(index);
+        oldValue = current.value;
+
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getCurrentNode(index);
+        Node<T> next = current.next;
+        Node<T> previous = current.previous;
+        final T deletedValue = current.value;
+
+        if (previous == null) {
+            head = next;
+        } else {
+            previous.next = next;
+            current.previous = null;
+        }
+
+        if (next == null) {
+            tail = previous;
+        } else {
+            next.previous = previous;
+            current.next = null;
+        }
+
+        current.value = null;
+        size--;
```

Comment:
let's move this code to the separate private method - unlink, and then we can use unlink method here and in the method remove(T object)